### PR TITLE
feat(data) create schematics to data

### PR DESCRIPTION
**Pull request by norato**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngrx/platform/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

There are no schematics to data

Closes #

## What is the new behavior?

Add schematics to create a data entity

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: docs/schematics/README.md [hunk 1]
```diff
@@ -87,6 +87,7 @@ ng generate @ngrx/schematics:effect App --root --module app.module.ts
 - [Feature](feature.md)
 - [Reducer](reducer.md)
 - [Store](store.md)
+- [Data](data.md)
```
Comment (timdeschryver) [inline docs/schematics/README.md]: You can remove this, NgRx data is for version 8 and up.

File: docs/schematics/data.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
+---
```
Comment (timdeschryver) [inline docs/schematics/data.md]: You can remove this, NgRx data is for version 8 and up.

File: modules/schematics/src/data/files/__name@dasherize@if-flat__/__name@dasherize__.ts.template [hunk 3]
```diff
@@ -0,0 +1,3 @@
+export interface <%= classify(name) %> {
+  id?: any;
```
Comment (timdeschryver) [inline modules/schematics/src/data/files/__name@dasherize@if-flat__/__name@dasherize__.ts.template]: Is having an `id` not a requirement?

```suggestion
  id: any;
```

File: modules/schematics/src/data/files/__name@dasherize@if-flat__/__name@dasherize__.service.spec.ts.template [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+import {
+  EntityCollectionServiceElementsFactory
+} from '@ngrx/data';
+import { <%= classify(name) %>Service } from './<%= dasherize(name) %>.service';
+
+describe('<%= classify(name) %>Service', () => {
+  let service: <%= classify(name) %>Service;
+
+  beforeEach(async () => {
+    TestBed.configureTestingModule({
```
Comment (timdeschryver) [inline modules/schematics/src/data/files/__name@dasherize@if-flat__/__name@dasherize__.service.spec.ts.template]:
```suggestion
    await TestBed.configureTestingModule({
```

Or to keep it consistent:

```
TestBed.configureTestingModule({
  providers: [
    EntityCollectionServiceElementsFactory,
    <%= classify(name) %>Service
  ]
});

await TestBed.compileComponents();
```

File: modules/schematics/src/data/index.spec.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+  const projectPath = getTestProjectPath();
+
+  let appTree: UnitTestTree;
+
+  beforeEach(async () => {
+    appTree = await createWorkspace(schematicRunner, appTree);
+  });
+
+  it('should create an data to specified project if provided', () => {
```
Comment (timdeschryver) [inline modules/schematics/src/data/index.spec.ts]:
```suggestion
  it('should create the data service to a specified project if provided', () => {
```

File: modules/schematics/src/data/index.spec.ts [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    const tree = schematicRunner.runSchematic('data', options, appTree);
+    expect(
+      tree.files.indexOf(`${projectPath}/src/app/foo.service.spec.ts`)
+    ).toBeGreaterThanOrEqual(0);
+    expect(
+      tree.files.indexOf(`${projectPath}/src/app/foo.service.ts`)
+    ).toBeGreaterThanOrEqual(0);
+  });
+
+  describe('service classe', () => {
```
Comment (timdeschryver) [inline modules/schematics/src/data/index.spec.ts]:
```suggestion
  describe('service class', () => {
```

File: modules/schematics/src/data/index.spec.ts [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+      const options = { ...defaultOptions, spec: true };
+      const tree = schematicRunner.runSchematic('data', options, appTree);
+      const fileContent = tree.readContent(
+        `${projectPath}/src/app/foo.service.ts`
+      );
+
+      expect(fileContent).toMatch(/import { Foo } from '.\/foo';/);
+    });
+
+    it('should extending EntityCollectionServiceBase', () => {
```
Comment (timdeschryver) [inline modules/schematics/src/data/index.spec.ts]: can we also test if `super` is called with the correct entity name?

File: modules/schematics/src/data/files/__name@dasherize@if-flat__/__name@dasherize__.ts.template (hunk 3 above)
Comment (norato) [reply modules/schematics/src/data/files/__name@dasherize@if-flat__/__name@dasherize__.ts.template]: It is just to suggest the interface to the developer. Setting it as required property on creating a new object in CRUD we do not have the id before receiving from backend it. But it is just a suggestion.

File: modules/schematics/src/data/index.spec.ts (hunk 6 above)
Comment (norato) [reply modules/schematics/src/data/index.spec.ts]: Portuguese, sorry

File: modules/schematics/src/data/schema.json [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+      "type": "string",
+      "$default": {
+        "$source": "argv",
+        "index": 0
+      }
+    },
+    "path": {
+      "type": "string",
+      "format": "path",
+      "description": "The path to create the component.",
```
Comment (timdeschryver) [inline modules/schematics/src/data/schema.json]:
```suggestion
      "description": "The path to create the service.",
```

File: modules/schematics/src/data/index.ts [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+  SchematicContext,
+  Tree,
+  url,
+} from '@angular-devkit/schematics';
+import {
+  getProjectPath,
+  parseName,
+  stringUtils,
+} from '@ngrx/schematics/schematics-core';
+import { Schema as ActionOptions } from './schema';
```
Comment (timdeschryver) [inline modules/schematics/src/data/index.ts]:
```suggestion
import { Schema as DataOptions } from './schema';
```

**Comment by timdeschryver** [comment]
@norato there also seems to be a conflict, could you resolve this please?

**Comment by timdeschryver** [comment]
@norato would it be possible to rebase to master please, so that only your new commits will be in this PR.

**Comment by norato** [comment]
@timdeschryver rebased

**Pull request merged**